util/src/main/java/org/datacommons/util/DcidGenerator.java

```diff
@@ -0,0 +1,369 @@
+package org.datacommons.util;
+
+import com.google.common.base.Charsets;
+import com.google.common.hash.Hashing;
+import java.nio.file.Path;
+import java.util.*;
+import org.datacommons.proto.Debug;
+import org.datacommons.proto.Mcf;
+
+// Helper functions to generate dcids for various class types.
+// Unless a node passes checks, its dcid cannot be generated. The functions in this class assume
+// those checks have already happened. So they do not report anything in LogWrapper. If a dcid
+// was not successfully generated, then a prior check will say why that is. This is to avoid
+// double counting failures.
+//
+// TODO: Have a way to assert that node whose dcid cannot be generated has an ERROR entry in
+//  LogWrapper.
+// TODO: Support hashed short-IDs for places. farmhash32 doesn't seem to exist in Java.
+public class DcidGenerator {
+  public static boolean TEST_MODE = false;
+
+  private static final String DC_NAMESPACE = "dc/";
+  private static final String OBS_NAMESPACE = "o/";
+  private static final String SVOBS_NAMESPACE = "o/";
+  private static final String POP_NAMESPACE = "p/";
+
+  // This is an int->char map of 32 chars.
+  private static final char[] DCID_BASE32_MAP = {
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'b', 'c', 'd', 'f', 'g', 'h', 'j', 'k', 'l',
+    'm', 'n', 'p', 'q', 'r', 's', 't', 'v', 'w', 'x', 'y', 'z', 'e'
+  };
+  // Max length of a base 32 encoded version of a 64-bit string: ceil(64 / 5)
+  private static final int MAX_LONG_ID_LEN = 13;
+  // Max length of a base 32 encoded version of a 32-bit string: ceil(32 / 5)
+  private static final int MAX_SHORT_ID_LEN = 7;
+  // Max number of bits of a 32-bit encoding
+  private static final int MAX_NUM_BITS = 5;
+
+  // Set of place types that use random short-ID. Typically though their IDs are derived from
+  // external IDs (like geoId/06).
+  private static final Set<String> TYPES_USING_SHORT_ID =
+      Set.of(
+          "City",
+          "County",
+          "Province",
+          "State",
+          "Country",
+          "Continent",
+          "CollegeOrUniversity",
+          "Source",
+          "Curator",
+          "Provenance",
+          "AdministrativeArea",
+          "AdministrativeArea1",
+          "AdministrativeArea2",
+          "AdministrativeArea3",
+          "AdministrativeArea4",
+          "AdministrativeArea5",
+          "Place",
+          "Town",
+          "Village",
+          "Neighborhood");
+
+  private static final Set<String> FIXED_STAT_VAR_PROPS =
+      Set.of(
+          Vocabulary.TYPE_OF,
+          Vocabulary.POPULATION_TYPE,
+          Vocabulary.MEASURED_PROP,
+          Vocabulary.STAT_TYPE,
+          Vocabulary.MEASUREMENT_QUALIFIER,
+          Vocabulary.MEASUREMENT_DENOMINATOR);
+
+  static final int LAST_REQUIRED_STAT_VAR_OBS_PROP_INDEX = 3; // until value
```

Comment:
i wonder if this should be defined after the list, and then the value derived with code (so that it's more robust to changes). something like below, but it's a bit verbose
```java
LAST_REQUIRED_STAT_VAR_OBS_PROP_INDEX = ORDERED_STAT_VAR_OBS_KEY_PROPS.indexOf(Vocabulary.GENERIC_VALUE)
```

It could also be the comment though? I needed to look twice to understand "until value"

Reply:
That is a neat idea!  Need to remember you can do such things with constants in java :)